Windows shared library build fails with LNK1104: cannot open file '..\glslang\Debug\glslang-default-resource-limitsd.lib'

I just tried a DLL build of glslang (passed `-DBUILD_SHARED_LIBS` to CMake), and hit three fatal errors. Two were `fatal error LNK1104: cannot open file '..\glslang\Debug\glslang-default-resource-limitsd.lib'` and one was `fatal error LNK1104: cannot open file '..\SPIRV\Debug\SPVRemapperd.lib'`. The files are specified with the `/IMPLIB` flag for the linker commands that would be expected to generate them.

I've had a look at the DLLs with this tool https://github.com/lucasg/Dependencies and it seems that they don't actually export any symbols. MSVC doesn't bother to emit an import library when a DLL exports no symbols. You can make GCC behave like MSVC here (and shrink binary size) by passing `-fvisibility=hidden` and marking everything as `__attribute__((__visibility__("default")))` if it'd be `__declspec(dllexport)` with MSVC.

It's possible that some `GLSLANG_EXPORT` macro calls are missing and symbols are supposed to be exported but aren't - after all, GCC and Clang both export all symbols by default, so it's easy to miss. If no functions are exported, linking with these DLLs won't do anything other than running their static initialisers and destructors, and none of their functions will be callable. Looking at the headers for these libraries, it certainly *looks* like there are functions that should be callable.

When this is fixed, I'd suggest adding a Windows DLL build to the CI so it doesn't get broken again.

Thanks, I will have fixes for the other stuff shortly, just need to do a
bit more testing before I push it.

On Fri, Sep 13, 2024, 07:08 Chris Djali ***@***.***> wrote:

> It at least seems to fix the problem where the import library isn't
> created at all thanks to nothing being exported, but there are still some
> errors about unresolved symbols when linking the tests and the standalone
> compiler. The tests complain that:
>
> 33>VkRelaxed.FromFile.obj : error LNK2019: unresolved external symbol "public: __cdecl glslang::TDefaultGlslIoResolver::TDefaultGlslIoResolver(class glslang::TIntermediate const &)" ***@***.***@@***@***.***@1@@Z) referenced in function "public: virtual void __cdecl glslangtest::`anonymous namespace'::VulkanRelaxedTest_FromFile_Test::TestBody(void)" ***@***.***_FromFile_Test@***@***.***@@UEAAXXZ)
> 33>GlslMapIO.FromFile.obj : error LNK2001: unresolved external symbol "public: __cdecl glslang::TDefaultGlslIoResolver::TDefaultGlslIoResolver(class glslang::TIntermediate const &)" ***@***.***@@***@***.***@1@@Z)
> 33>VkRelaxed.FromFile.obj : error LNK2019: unresolved external symbol "public: __cdecl glslang::TGlslIoMapper::TGlslIoMapper(void)" ***@***.***@@***@***.***) referenced in function "public: virtual void __cdecl glslangtest::`anonymous namespace'::VulkanRelaxedTest_FromFile_Test::TestBody(void)" ***@***.***_FromFile_Test@***@***.***@@UEAAXXZ)
> 33>GlslMapIO.FromFile.obj : error LNK2001: unresolved external symbol "public: __cdecl glslang::TGlslIoMapper::TGlslIoMapper(void)" ***@***.***@@***@***.***)
> 33>VkRelaxed.FromFile.obj : error LNK2019: unresolved external symbol "public: virtual __cdecl glslang::TGlslIoMapper::~TGlslIoMapper(void)" ***@***.***@@***@***.***) referenced in function "public: virtual void __cdecl glslangtest::`anonymous namespace'::VulkanRelaxedTest_FromFile_Test::TestBody(void)" ***@***.***_FromFile_Test@***@***.***@@UEAAXXZ)
> 33>GlslMapIO.FromFile.obj : error LNK2001: unresolved external symbol "public: virtual __cdecl glslang::TGlslIoMapper::~TGlslIoMapper(void)" ***@***.***@@***@***.***)
> 33>VkRelaxed.FromFile.obj : error LNK2019: unresolved external symbol "public: bool __cdecl glslang::TReflection::addStage(enum EShLanguage,class glslang::TIntermediate const &)" ***@***.***@glslang@@QEAA_NW4EShLanguage@@***@***.***@@Z) referenced in function "bool __cdecl glslangtest::`anonymous namespace'::verifyIOMapping(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > &,class glslang::TProgram &)" (?verifyIOMapping@***@***.***@@***@***.******@***.***@std@@***@***.***@2@@std@@***@***.***@@@Z)
> 33>GlslMapIO.FromFile.obj : error LNK2001: unresolved external symbol "public: bool __cdecl glslang::TReflection::addStage(enum EShLanguage,class glslang::TIntermediate const &)" ***@***.***@glslang@@QEAA_NW4EShLanguage@@***@***.***@@Z)
> 33>GlslMapIO.FromFile.obj : error LNK2019: unresolved external symbol "public: static int __cdecl glslang::TIntermediate::computeTypeLocationSize(class glslang::TType const &,enum EShLanguage)" ***@***.***@glslang@@***@***.***@W4EShLanguage@@@Z) referenced in function "bool __cdecl glslangtest::`anonymous namespace'::verifyIOMapping(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > &,class glslang::TProgram &)" (?verifyIOMapping@***@***.***@@***@***.******@***.***@std@@***@***.***@2@@std@@***@***.***@@@Z)
> 33>C:\vsg\glslang\build\MSVC2022-dynamic\gtests\Debug\glslangtests.exe : fatal error LNK1120: 5 unresolved externals
>
> and the standalone compiler complains that
>
> 34>StandAlone.obj : error LNK2019: unresolved external symbol "void __cdecl glslang::OS_DumpMemoryCounters(void)" ***@***.***@@YAXXZ) referenced in function "void __cdecl CompileFile(char const *,void *)" (?CompileFile@@***@***.***)
> 34>C:\vsg\glslang\build\MSVC2022-dynamic\StandAlone\Debug\glslang.exe : fatal error LNK1120: 1 unresolved externals
>
> Adding GLSLANG_EXPORT to the symbols declared in osinclude.h and
> iomapper.h resolve these *except* for a couple of functions:
>
> 20>glslang_c_interface.obj : error LNK2019: unresolved external symbol "protected: __cdecl glslang::TDefaultIoResolverBase::TDefaultIoResolverBase(struct glslang::TDefaultIoResolverBase &)" ***@***.***@@***@***.***@@Z) referenced in function "public: __cdecl glslang::TDefaultGlslIoResolver::TDefaultGlslIoResolver(struct glslang::TDefaultGlslIoResolver &&)" ***@***.***@@QEAA@$$QEAU01@@Z)
>
> This one genuinely doesn't seem to have an implementation, even though
> it's declared. GCC would just not export the TDefaultGlslIoResolver move
> constructor as it's not called, and without the export declaration, MSVC
> would have skipped over it, too.
>
> 20>glslang_c_interface.obj : error LNK2019: unresolved external symbol "protected: struct glslang::TDefaultIoResolverBase & __cdecl glslang::TDefaultIoResolverBase::operator=(struct glslang::TDefaultIoResolverBase &)" ***@***.***@@***@***.***@@Z) referenced in function "public: struct glslang::TDefaultGlslIoResolver & __cdecl glslang::TDefaultGlslIoResolver::operator=(struct glslang::TDefaultGlslIoResolver &&)" ***@***.***@@QEAAAEAU01@$$QEAU01@@Z)
>
> The same as above, it's declared but not implemented.
>
> 22>VkRelaxed.FromFile.obj : error LNK2019: unresolved external symbol "public: bool __cdecl glslang::TReflection::addStage(enum EShLanguage,class glslang::TIntermediate const &)" ***@***.***@glslang@@QEAA_NW4EShLanguage@@***@***.***@@Z) referenced in function "bool __cdecl glslangtest::`anonymous namespace'::verifyIOMapping(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > &,class glslang::TProgram &)" (?verifyIOMapping@***@***.***@@***@***.******@***.***@std@@***@***.***@2@@std@@***@***.***@@@Z)
>
> Fixed by marking exports in reflection.h.
>
> 22>GlslMapIO.FromFile.obj : error LNK2019: unresolved external symbol "public: static int __cdecl glslang::TIntermediate::computeTypeLocationSize(class glslang::TType const &,enum EShLanguage)" ***@***.***@glslang@@***@***.***@W4EShLanguage@@@Z) referenced in function "bool __cdecl glslangtest::`anonymous namespace'::verifyIOMapping(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > &,class glslang::TProgram &)" (?verifyIOMapping@***@***.***@@***@***.******@***.***@std@@***@***.***@2@@std@@***@***.***@@@Z)
>
> Fixed by marking exports in localintermediate.h.
>
> I've pushed my extra exports to
> https://github.com/AnyOldName3/glslang/tree/spvremapper-visibility. I've
> not done anything about the declared-but-not-implemented functions as it's
> not clear whether they were a failed attempt to make the type non-copyable
> or whether they were supposed to be implemented.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/KhronosGroup/glslang/issues/3716#issuecomment-2349051020>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A5I72B76B4PZ2ENU7UOBSFLZWLWUHAVCNFSM6AAAAABNTAZPISVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDGNBZGA2TCMBSGA>
> .
> You are receiving this because you were assigned.Message ID:
> ***@***.***>
>

--- a/SPIRV/CMakeLists.txt
+++ b/SPIRV/CMakeLists.txt
@@ -100,6 +100,7 @@ if (ENABLE_SPVREMAPPER)
         POSITION_INDEPENDENT_CODE ON
         VERSION   "${GLSLANG_VERSION}"
         SOVERSION "${GLSLANG_VERSION_MAJOR}")
+    glslang_only_export_explicit_symbols(SPVRemapper)
 endif()
 
 if(WIN32 AND BUILD_SHARED_LIBS)

--- a/SPIRV/SPVRemapper.h
+++ b/SPIRV/SPVRemapper.h
@@ -41,6 +41,21 @@
 #include <cstdlib>
 #include <exception>
 
+#ifdef GLSLANG_IS_SHARED_LIBRARY
+    #ifdef _WIN32
+        #ifdef GLSLANG_EXPORTING
+            #define GLSLANG_EXPORT __declspec(dllexport)
+        #else
+            #define GLSLANG_EXPORT __declspec(dllimport)
+        #endif
+    #elif __GNUC__ >= 4
+        #define GLSLANG_EXPORT __attribute__((visibility("default")))
+    #endif
+#endif // GLSLANG_IS_SHARED_LIBRARY
+#ifndef GLSLANG_EXPORT
+#define GLSLANG_EXPORT
+#endif
+
 namespace spv {
 
 class spirvbin_base_t
@@ -83,7 +98,7 @@ namespace spv {
 static inline constexpr Id NoResult = 0;
 
 // class to hold SPIR-V binary data for remapping, DCE, and debug stripping
-class spirvbin_t : public spirvbin_base_t
+class GLSLANG_EXPORT spirvbin_t : public spirvbin_base_t
 {
 public:
    spirvbin_t(int verbose = 0) : entryPoint(spv::NoResult), largestNewId(0), verbose(verbose), errorLatch(false)

